pkg/cache/v2/simple_delta.go

```diff
@@ -0,0 +1,297 @@
+					unsubscribed,
+					version,
+				)
+
+				// discard the old watch
+				delete(info.deltaWatches, id)
+			} else if version != info.deltaState[t].Version {
+				if len(subscribed) == 0 && len(unsubscribed) == 0 {
+					cache.log.Debugf("wildcard request")
+
+					// we should set our delta state here somehow
+					// Maybe set the resources for all the types here???
+					for i := 0; i < int(types.UnknownType); i++ {
+						tURL := GetResponseTypeURL(types.ResponseType(i))
+						info.deltaState[tURL] = Resources{
+							Version: version,
+							Items:   snapshot.GetResources(tURL),
+						}
+					}
+				}
+
+				// Assume we've received a new resource and we want to send new resources and cancel old watches
+				diff := cache.checkState(subscribed, info.deltaState[t].Items)
+				if len(diff) > 0 {
+					if cache.log != nil {
+						cache.log.Debugf("node: %s, found new items to subscribe too: %v ", node, diff)
+					}
+
+					// Add our new subscription items to our state to watch that we've found
+					r := Resources{
+						Version: version,
+					}
+					for key, value := range diff {
+						for rKey := range info.deltaState[t].Items {
+							// Handle the case when a new item could be added to the state and also if we need to overwrite a previous resource
+							if key == rKey {
+								info.deltaState[t].Items[key] = value
+							} else if _, found := info.deltaState[t].Items[key]; !found {
+								info.deltaState[t].Items[key] = value
+							}
+						}
+					}
+					r.Items = info.deltaState[t].Items
+
+					info.deltaState[t] = r
+				}
+
+				if len(unsubscribed) > 0 {
+					// we need to remove the previously subscribed resources from the state so we no longer send updates
+					if cache.log != nil {
+						cache.log.Debugf("node: %s, recieved items to unsubscribe from: %v", node, unsubscribed)
+					}
+
+					newState := cache.unsubscribe(unsubscribed, info.deltaState)
+					for typ := range info.deltaState {
+						info.deltaState[typ] = newState[typ]
+					}
+				}
+
+				if cache.log != nil {
+					// We only want to show the specific resources we're sending back from the diff
+					cache.log.Debugf("delta respond open watch ID:%d Resources:%+v with new version %q", id, diff, version)
+				}
+
+				// Respond to our delta stream with the subcribed resources
+				cache.respondDelta(
+					watch.Request,
+					watch.Response,
+					diff, // We want to only send the updated resources here
+					unsubscribed,
+					version,
+				)
+
+				// discard the old watch
+				delete(info.deltaWatches, id)
+			}
+		}
+		info.mu.Unlock()
+	}
+
+	return nil
+}
+
+// difference returns the elements in `a` that aren't in `b`.
+// TODO: SLOW this will need to be revisited
+func (cache *snapshotCache) checkState(resources, deltaState map[string]types.Resource) map[string]types.Resource {
+	mb := make(map[string]types.Resource, len(deltaState))
+	diff := make(map[string]types.Resource, 0)
+
+	for key, value := range deltaState {
```

Comment:
Is there a reason you are cloning the `deltaState`?  You don't mutate `mb` below so the only reason is if you think `deltaState` is mutated in another thread.

Reply:
If I don't copy it here it could get mutated by another thread or stream. I originally tried to not copy it and was seeing some odd behavior like things missing or not being detected.

Reply:
Hmm, and now it looks like it's not being used at all below?

Reply:
Ah yes thank you, I was confused for a second but you're right, it wasn't being used `mb` was basically useless. Deleted.